extensions/amp-app-banner/1.0/component/android.js

```diff
@@ -0,0 +1,104 @@
+import {docInfoService} from '#preact/services/document';
+import {platformService} from '#preact/services/platform';
+import {timerService} from '#preact/services/timer';
+import {urlService} from '#preact/services/url';
+import {viewerService} from '#preact/services/viewer';
+import {xhrService} from '#preact/services/xhr';
+
+import {user} from '#utils/log';
+
+import {openWindowDialog} from '../../../../src/open-window-dialog';
+
+const OPEN_LINK_TIMEOUT = 1500;
+
+/**
+ * @return {{openOrInstall: function(): void, promise: Promise<Response>}|null}
+ */
+export function getAndroidAppInfo() {
+  const win = self.window;
+  // We want to fallback to browser builtin mechanism when possible.
+  const isChromeAndroid =
+    platformService.isAndroid() && platformService.isChrome();
+  const isProxyOrigin = urlService.isProxyOrigin(win.location);
+  const canShowBuiltinBanner =
+    !isProxyOrigin && !viewerService.isEmbedded() && isChromeAndroid;
+
+  if (canShowBuiltinBanner) {
+    user().info(
+      'BENTO-APP-BANNER',
+      'Browser supports builtin banners. Not rendering amp-app-banner.'
+    );
+    return null;
+  }
+
+  const manifestLink = win.document.head.querySelector(
+    'link[rel=manifest],link[rel=origin-manifest]'
+  );
+
+  const missingDataSources = !manifestLink;
+  if (missingDataSources) {
+    return null;
+  }
+
+  const manifestHref = manifestLink.getAttribute('href');
+
+  urlService.assertHttpsUrl(manifestHref, undefined, 'manifest href');
+
+  const promise = xhrService.fetchJson(manifestHref).then(parseManifest);
+  return {
+    promise,
+    openOrInstall: () => {
+      return promise.then((manifest) => {
+        if (!manifest) {
+          return;
+        }
+        const {installAppUrl, openInAppUrl} = manifest;
+        timerService.delay(() => {
+          window.top.location.assign(installAppUrl);
```

Comment:
This code was copied from the Amp Classic implementation ... Should I change it in both places, to keep it "symmetric"?

And yes, I agree, this code was hard to test, I had to stub `timerService.delay` to bypass this from getting called, because I couldn't stub any part of `window.top.location.assign`.